content/blog/example-of-mapreduce-with-ezmeral-data-fabric-mapr-database-binary-table.md

```diff
@@ -0,0 +1,365 @@
+
+
+**It's worth noting that installing Docker Desktop in Windows won't work.**
+
+
+I tried the following: first install WSL2 (Windows Subsystem Linux 2), then install Docker Desktop for Windows and integrate with WSL2, then run the EDF Development Environment install script, but it still fails.
+
+
+So what I ended up doing was: install VMWare on my Windows PC, create a CentOS8 VM, and run the EDF Development Environment setup script in the VM. This approach is feasible.
+Also, you can always choose the version of the Development Environment you want to deploy, you just need to change the tag of the Docker image.
+
+
+
+## MapReduce on EDF Database Binary Table
+
+
+
+EDF Database Binary Table is equivalent to the EDF version of Apache HBase, but its technical implementation is different from HBase, which is of course, because the bottom layer of EDF Database Binary Table is EDF File Store.
+
+For users, there is almost no difference between using EDF Database Binary Table and using HBase.
+
+
+Now, let's imagine that we want to build a User Notifications service.
+
+
+Since HBase does not support any operations that across rows or across tables, in order to implement operations such as Joins and Group by in RDBMS, we will use MapReduce to complete some data analysis tasks.
+
+
+
+### Create a Binary Table
+
+
+
+**Important note**:
+
+<ins>the commands in this article are all executed as the user "mapr", which is by default the admin user of the EDF Development Environment.</ins>
+
+<ins>You can also use the "root" user to create the table and run the application, but if you don't modify the ACEs of the table, the "mapr" user would be not able to see the data in the table.</ins>
+
+
+
+We are going to use the [hbase shell][HbaseShell] to create a [Binary Table][BinaryTable] inside the EDF Database.
+To be able to use the `hbase shell`, we need to install the **[mapr-hbase][MaprHbasePackage]** package first.
+
+
+
+#### For convenience of data management, I would like to create a volume for the Binary Table, the command are as following:
+
+
+
+```bash
+sudo -u mapr maprcli volume list -output terse -columns volumename,volumetype,actualreplication,localpath,mounted,mountdir,logicalUsed,used,nameContainerDataThresholdMB,nameContainerSizeMB,needsGfsck
+
+maprcli volume create -name test.binarytable1 -path /testbinarytable1volume
+
+# sudo -u mapr maprcli volume mount -name test.binarytable1 -path /testbinarytable1volume
+sudo -u mapr hadoop fs -ls -d -h /testbinarytable1volume
+sudo -u mapr hadoop mfs -ls /testbinarytable1volume
+```
+
+
+
+☝ The volume's name is: test.binarytable1 and it will be mounted as <ins>/testbinarytable1volume/</ins> in the EDF File System.
+
+
+
+#### Now we can create the Binary Table
```

Comment:
I would suggest to use bold characters (**) instead of Heading level 4 throughout the blog post.